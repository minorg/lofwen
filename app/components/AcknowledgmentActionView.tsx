import { View } from "react-native";
import {} from "~/components/ui/radio-group";
import { Text } from "~/components/ui/text";
import type { AcknowledgmentAction, Event } from "~/models";

export function AcknowledgmentActionView({
  action,
}: {
  action: AcknowledgmentAction;
  onEvent: (event: Event) => void;
}) {
  return (
    <View className="flex flex-col flex-1 native:justify-center native:px-4">
      <Text className="text-2xl">{action.message}</Text>
    </View>
  );
}
