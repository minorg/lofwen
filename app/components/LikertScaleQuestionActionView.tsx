import { useCallback, useMemo } from "react";
import { View } from "react-native";
import { Label } from "~/components/ui/label";
import { RadioGroup, RadioGroupItem } from "~/components/ui/radio-group";
import { Text } from "~/components/ui/text";
import { useLog } from "~/hooks/useLog";
import { logger } from "~/logger";
import type { Event, LikertScaleQuestionAction } from "~/models";

export function LikertScaleQuestionActionView({
  action: question,
  onEvent,
}: {
  action: LikertScaleQuestionAction;
  onEvent: (event: Event) => Promise<void>;
}) {
  const log = useLog();
  const answer = useMemo(() => {
    const answer = log.answerEvent(question);
    if (answer !== null) {
      logger.debug(
        "existing answer to question",
        question["@id"],
        ":",
        JSON.stringify(answer),
      );
    } else {
      logger.debug("no existing answer to question", question["@id"]);
    }
    return answer;
  }, [log, question]);

  const onSelectResponseCategoryLabel = useCallback(
    (responseCategoryLabel: string) =>
      onEvent({
        "@type": "LikertScaleAnswerEvent",
        questionActionId: question["@id"],
        responseCategory: question.responseCategories.find(
          (responseCategory) =>
            responseCategory.label === responseCategoryLabel,
        )!,
      }),
    [onEvent, question],
  );

  return (
    <View className="flex flex-col gap-2 native:justify-center native:px-4">
      <Text className="text-2xl">{question.prompt}</Text>
      <RadioGroup
        onValueChange={onSelectResponseCategoryLabel}
        value={answer?.responseCategory.label ?? ""}
      >
        {question.responseCategories.map((responseCategory) => (
          <View
            className={"flex-row items-center gap-2"}
            key={responseCategory.label}
          >
            <RadioGroupItem
              aria-labelledby={`label-for-${responseCategory.label}`}
              value={responseCategory.label}
            />
            <Label
              nativeID={`label-for-${responseCategory.label}`}
              onPress={() =>
                onSelectResponseCategoryLabel(responseCategory.label)
              }
            >
              <Text className="text-xl">{responseCategory.label}</Text>
            </Label>
          </View>
        ))}
      </RadioGroup>
    </View>
  );
}
