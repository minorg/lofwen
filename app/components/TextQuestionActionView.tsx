import { useCallback, useEffect, useState } from "react";
import { View } from "react-native";
import { Button } from "~/components/ui/button";
import { Input } from "~/components/ui/input";
import {} from "~/components/ui/radio-group";
import { Text } from "~/components/ui/text";
import { useLog } from "~/hooks/useLog";
import { logger } from "~/logger";
import type { Event, TextQuestionAction } from "~/models";

export function TextQuestionActionView({
  action: question,
  onEvent,
}: {
  action: TextQuestionAction;
  onEvent: (event: Event) => Promise<void>;
}) {
  const log = useLog();
  const [text, setText] = useState("");

  useEffect(() => {
    const answer = log.answerEvent(question);
    if (answer !== null) {
      setText(answer.text);
    } else {
      logger.debug("no existing answer to question", question["@id"]);
    }
  }, [log, question]);

  const onSubmitButtonPress = useCallback(
    () =>
      onEvent({
        "@type": "TextAnswerEvent",
        questionActionId: question["@id"],
        text,
      }),
    [onEvent, question, text],
  );

  return (
    <View className="flex flex-col flex-1 gap-2 native:justify-center native:px-4">
      <Text className="text-2xl">{question.prompt}</Text>
      <Input
        className="h-[8rem]"
        multiline
        onChangeText={setText}
        value={text}
      />
      <Button
        disabled={text.length === 0}
        onPress={onSubmitButtonPress}
        variant="outline"
      >
        <Text>Submit</Text>
      </Button>
    </View>
  );
}
