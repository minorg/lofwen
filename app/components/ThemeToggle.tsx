import { useCallback } from "react";
import { Pressable, View } from "react-native";
import { useColorScheme } from "~/hooks/useColorScheme";
import { MoonStar } from "~/icons/MoonStar";
import { Sun } from "~/icons/Sun";
import { setAndroidNavigationBar } from "~/lib/setAndroidNavigationBar";
import { cn } from "~/lib/utils";

export function ThemeToggle() {
  const { isDarkColorScheme, setColorScheme } = useColorScheme();

  const toggleColorScheme = useCallback(() => {
    const newTheme = isDarkColorScheme ? "light" : "dark";
    setAndroidNavigationBar(newTheme);
    setColorScheme(newTheme);
  }, [isDarkColorScheme, setColorScheme]);

  return (
    <Pressable
      onPress={toggleColorScheme}
      className="web:ring-offset-background web:transition-colors web:focus-visible:outline-none web:focus-visible:ring-2 web:focus-visible:ring-ring web:focus-visible:ring-offset-2"
    >
      {({ pressed }) => (
        <View
          className={cn(
            "flex-1 aspect-square pt-0.5 justify-center items-start web:px-5",
            pressed && "opacity-70",
          )}
        >
          {isDarkColorScheme ? (
            <MoonStar
              className="text-foreground"
              size={23}
              strokeWidth={1.25}
            />
          ) : (
            <Sun className="text-foreground" size={24} strokeWidth={1.25} />
          )}
        </View>
      )}
    </Pressable>
  );
}
