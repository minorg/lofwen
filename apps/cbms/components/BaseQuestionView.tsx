import type { PropsWithChildren } from "react";
import { View } from "react-native";
import { Text } from "~/components/ui/text";
import type { Question } from "~/models/Question";

export function BaseQuestionView({
  children,
  question,
}: PropsWithChildren<{ question: Question }>) {
  return (
    <View className="flex flex-col gap-2 native:justify-center native:p-4">
      <Text className="text-primary">{question.prompt}</Text>
      {children}
    </View>
  );
}
