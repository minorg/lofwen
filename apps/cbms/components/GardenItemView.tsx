import { Pressable, View } from "react-native";
import { IconView } from "~/components/IconView";
import { Button } from "~/components/ui/button";
import { Text } from "~/components/ui/text";
import { useColorScheme } from "~/hooks/useColorScheme";
import type { GardenItem } from "~/models/GardenItem";

export function GardenItemView({
  gardenItem,
  onSelect,
}: { gardenItem: GardenItem; onSelect: () => void }) {
  const { isDarkColorScheme } = useColorScheme();

  return (
    <View className="border-2 border-primary flex flex-row gap-4 p-4 rounded-sm">
      <View className="flex flex-col justify-center">
        <Pressable onPress={onSelect}>
          <IconView
            color={isDarkColorScheme ? "white" : undefined}
            icon={gardenItem.icon}
            size={64}
          />
        </Pressable>
      </View>
      <View className="flex flex-col flex-1 items-end">
        <Button onPress={onSelect} variant="outline">
          <Text className="text-lg font-bold">{gardenItem.title}</Text>
        </Button>
        <Text className="text-secondary">{gardenItem.text}</Text>
      </View>
    </View>
  );
}
