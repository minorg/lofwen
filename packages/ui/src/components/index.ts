export * from "./ui/button";
export * from "./ui/input";
export * from "./ui/label";
export * from "./ui/radio-group";
export * from "./ui/text";
export * from "./LikertScaleQuestionView";
export * from "./TextQuestionView";
export * from "./ThemeToggle";
