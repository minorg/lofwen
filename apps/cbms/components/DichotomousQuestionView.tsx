import type { DichotomousAnswer, DichotomousQuestion } from "@lofwen/models";
import { useCallback } from "react";
import { View } from "react-native";
import { BaseQuestionView } from "~/components/BaseQuestionView";
import { Label } from "~/components/ui/label";
import { RadioGroup, RadioGroupItem } from "~/components/ui/radio-group";
import { Text } from "~/components/ui/text";

export function DichotomousQuestionView({
  answer,
  onAnswer,
  question,
}: {
  answer: DichotomousAnswer | null;
  onAnswer: (answer: DichotomousAnswer) => void;
  question: DichotomousQuestion;
}) {
  const onSelectResponseCategoryLabel = useCallback(
    (responseCategoryLabel: string) =>
      onAnswer({
        "@type": "DichotomousAnswer",
        responseCategory: question.responseCategories.find(
          (responseCategory) =>
            responseCategory.label === responseCategoryLabel,
        )!,
      }),
    [onAnswer, question],
  );

  return (
    <BaseQuestionView question={question}>
      <RadioGroup
        onValueChange={onSelectResponseCategoryLabel}
        value={answer?.responseCategory.label ?? ""}
      >
        {question.responseCategories.map((responseCategory) => (
          <View
            className={"flex-row items-center gap-2"}
            key={responseCategory.label}
          >
            <RadioGroupItem
              aria-labelledby={`label-for-${responseCategory.label}`}
              value={responseCategory.label}
            />
            <Label
              nativeID={`label-for-${responseCategory.label}`}
              onPress={() =>
                onSelectResponseCategoryLabel(responseCategory.label)
              }
            >
              <Text className="text-secondary">{responseCategory.label}</Text>
            </Label>
          </View>
        ))}
      </RadioGroup>
    </BaseQuestionView>
  );
}
