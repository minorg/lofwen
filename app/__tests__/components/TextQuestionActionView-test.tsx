import { render } from "@testing-library/react-native";
import { textQuestionAction } from "~/__tests__/data";
import { TextQuestionActionView } from "~/components/TextQuestionActionView";

it("renders correctly", () => {
  const tree = render(
    <TextQuestionActionView
      action={textQuestionAction}
      onEvent={async () => {}}
    />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
