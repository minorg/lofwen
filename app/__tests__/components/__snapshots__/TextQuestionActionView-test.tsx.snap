// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  className="flex flex-col flex-1 gap-2 native:justify-center native:px-4"
>
  <Text
    className="text-foreground web:select-text text-2xl"
  >
    Tell us what you like about the app.
  </Text>
  <TextInput
    className="web:flex native:h-12 web:w-full rounded-md border border-input bg-background px-3 web:py-2 text-base lg:text-sm native:text-lg native:leading-[1.25] text-foreground placeholder:text-muted-foreground web:ring-offset-background file:border-0 file:bg-transparent file:font-medium web:focus-visible:outline-none web:focus-visible:ring-2 web:focus-visible:ring-ring web:focus-visible:ring-offset-2 h-[8rem]"
    multiline={true}
    onChangeText={[Function]}
    placeholderClassName="text-muted-foreground"
    value=""
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    className="opacity-50 web:pointer-events-none group flex items-center justify-center rounded-md web:ring-offset-background web:transition-colors web:focus-visible:outline-none web:focus-visible:ring-2 web:focus-visible:ring-ring web:focus-visible:ring-offset-2 border border-input bg-background web:hover:bg-accent web:hover:text-accent-foreground active:bg-accent h-10 px-4 py-2 native:h-12 native:px-5 native:py-3"
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    role="button"
  >
    <Text
      className="web:select-text web:whitespace-nowrap text-sm native:text-base font-medium text-foreground web:transition-colors group-active:text-accent-foreground web:pointer-events-none"
    >
      Submit
    </Text>
  </View>
</View>
`;
